File: Daily/PY/Leetcode1723-完成所有工作的最短时间.py
from typing import List
class Solution:
    def __init__(self):
        self.jobs = []
        self.k = 0
    def minimumTimeRequired(self, jobs: List[int], k: int) -> int:
        self.jobs = jobs
        self.k = k
        l,r = max(jobs),sum(jobs)
        def check(mid):
            workloads = [0] * self.k
            return back(workloads,mid,0)
        def back(workloads,limit,idx):
            if idx == len(self.jobs):
                return True
            cur = self.jobs[idx]
            for i,_ in enumerate(workloads):
                if workloads[i] + cur <= limit:
                    workloads[i] += cur
                    if back(workloads, limit, idx + 1):
                        return True
                    workloads[i] -= cur
                if workloads[i] == 0 or workloads[i] + cur == limit:
                    break
            return False
        while l < r:
            mid = l + r >> 1
            if check(mid):
                r = mid
            else:
                l = mid + 1
        return l

File: Leetcode/Rust/Leetcode1723-完成所有工作的最短时间.rs
impl Solution {
    pub fn minimum_time_required(jobs: Vec<i32>, k: i32) -> i32 {
        if jobs.len()==0{
            return 0;
        }
        let mut beg = jobs.iter().max().unwrap().clone();
        let mut end:i32 = jobs.iter().sum();
        let k = k as usize;
        if jobs.len()==k{
            return beg;
        }
        let mut mid = (beg + end)>>1;
        while beg < end {
            mid = (beg + end)>>1;
            let mut work_times = vec![0;k ];
            if Self::dfs(0, &jobs[..], mid, k , &mut work_times[..]){
                end = mid;
            }else{
                beg = mid+1;
            }
        }
        beg

    }
    fn dfs(curjob_id: usize, jobs: &[i32], mid: i32, k: usize, work_times: &mut [i32]) -> bool {
        if curjob_id == jobs.len(){
            return true;
        }
        for worker in 0..k {
            if work_times[worker] + jobs[curjob_id] <= mid {
                work_times[worker] += jobs[curjob_id];
                if Self::dfs(curjob_id + 1, jobs, mid, k, work_times) {
                    return true;
                }

                work_times[worker] -= jobs[curjob_id];
            }

            if work_times[worker] == 0 {
                return false;
            }
        }
        false
    }
}